br/pkg/lightning/backend/local/region_job.go

```diff
@@ -0,0 +1,620 @@
+
+const (
+	regionScanned jobStageTp = "regionScanned"
+	wrote         jobStageTp = "wrote"
+	ingested      jobStageTp = "ingested"
+	needRescan    jobStageTp = "needRescan"
+)
+
+func (j jobStageTp) String() string {
+	return string(j)
+}
+
+// regionJob is dedicated to import the data in [keyRange.start, keyRange.end) to a region.
+type regionJob struct {
+	keyRange Range
+	// TODO: check the keyRange so that it's always included in region
+	region *split.RegionInfo
+	// stage should be updated only by convertStageTo
+	stage jobStageTp
+
+	engine          *Engine
+	regionSplitSize int64
+	regionSplitKeys int64
+	metrics         *metric.Metrics
+	// below fields are available after wrote stage
+	writeResult *tikvWriteResult
+
+	retryCount       int
+	waitUntil        time.Time
+	lastRetryableErr error
+}
+
+type tikvWriteResult struct {
+	sstMeta    []*sst.SSTMeta
+	rangeStats rangeStats
+}
+
+type rangeStats struct {
+	count      int64
+	totalBytes int64
+}
+
+func (j *regionJob) convertStageTo(stage jobStageTp) {
+	j.stage = stage
+	switch stage {
+	case regionScanned:
+		j.writeResult = nil
+	case ingested:
+		j.engine.finishedRanges.add(j.keyRange)
+
+		// when writing is skipped because range is empty
+		if j.writeResult == nil {
+			return
+		}
+
+		j.engine.importedKVSize.Add(j.writeResult.rangeStats.totalBytes)
+		j.engine.importedKVCount.Add(j.writeResult.rangeStats.count)
+		if j.metrics != nil {
+			j.metrics.BytesCounter.WithLabelValues(metric.BytesStateImported).
+				Add(float64(j.writeResult.rangeStats.totalBytes))
+		}
+	}
+}
+
+// writeToTiKV writes the data to TiKV and mark this job as wrote stage.
+// if any write logic has error, writeToTiKV will set job to a proper stage and return nil.
+// if any underlying logic has error, writeToTiKV will return an error.
+// we don't need to do cleanup for the pairs written to tikv if encounters an error,
+// tikv will take the responsibility to do so.
+// TODO: let client-go provide a high-level write interface.
+func (j *regionJob) writeToTiKV(
+	ctx context.Context,
+	clientFactory ImportClientFactory,
+	kvBatchSize int,
+	bufferPool *membuf.Pool,
+	writeLimiter StoreWriteLimiter,
+) error {
+	if j.stage != regionScanned {
+		return nil
+	}
+
+	begin := time.Now()
+	stats := rangeStats{}
+
+	firstKey, lastKey, err := j.engine.getFirstAndLastKey(j.keyRange.start, j.keyRange.end)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if firstKey == nil {
+		j.convertStageTo(ingested)
+		log.FromContext(ctx).Info("keys within region is empty, skip doIngest",
+			logutil.Key("start", j.keyRange.start),
+			logutil.Key("regionStart", j.region.Region.StartKey),
+			logutil.Key("end", j.keyRange.end),
+			logutil.Key("regionEnd", j.region.Region.EndKey))
+		return nil
+	}
+
+	firstKey = codec.EncodeBytes([]byte{}, firstKey)
+	lastKey = codec.EncodeBytes([]byte{}, lastKey)
+
+	u := uuid.New()
+	meta := &sst.SSTMeta{
+		Uuid:        u[:],
+		RegionId:    j.region.Region.GetId(),
```

Comment:
maybe make `j.region.Region` a local var, accessed a lot of times

Reply:
https://github.com/pingcap/tidb/pull/40692/commits/5479d517b1a5602817a5b3c84c5a495c3eda4e9f